domain/domain.go

```diff
@@ -199,6 +200,20 @@ func (do *Domain) loadInfoSchema(startTS uint64) (infoschema.InfoSchema, bool, i
 	if err != nil {
 		return nil, false, 0, nil, err
 	}
+	// fetch the commit timestamp of the schema diff
+	schemaTs := int64(0)
+	tikvStore, ok := do.Store().(helper.Storage)
+	if ok {
+		helper := helper.NewHelper(tikvStore)
```

Comment:
@xhebox thanks for the suggestion! i feel like it is perhaps better to use the mvcc commit ts of the schema diff as the timestamp, because add a new field in schema diff will need to deal with backward compatible that some schema may not have the timestamp while some other does. however, in order to fetch the mvcc, i am constructing a new `helper` like this, is this ok or it could / should be done better? thanks!

Reply:
> because add a new field in schema diff will need to deal with backward compatible that some schema may not have the timestamp while some other does. 

It is perfectly fine. We use `json`, that means additional fields won't break old clusters.

**And honestly. your current solution will also fail on old clusters, since old cluster have no corresponding meta info.**

You need to solve the problem that old clusters does not have a infoschema. It is only safe to execute the code after all tidb fetched the latest infoschema.

That is there should be no DDL while upgrading, which is not required by the current tidb. You either need a global system variable, or just do not cache infoschema with invalid TS, except the `maxSnapshotTS` one.

> however, in order to fetch the mvcc, i am constructing a new helper like this, is this ok or it could / should be done better

LGTM, this looks like the correct solution however it is backward compatibility breaker now.

Reply:
thanks for the review! i am addressing other comments, but have some confusion on the compatibility issue you mentioned? the timestamp used here is the mvcc commit timestamp of schema diff, so your concern is maybe for some historical version, the commit timestamp could not be available?

Reply:
> so your concern is maybe for some historical version, the commit timestamp could not be available?

Exactly, it is fine on TiDB cloud cluster. Our manager can handle that. But for self-managed clusters. it will be an unexpected compatibility breaker.

It is better to have a system variable to control, I think.

Reply:
> the timestamp used here is the mvcc commit timestamp of schema diff,

I am saying you can store `startTS` into schema diff, and get `commit TS` in `domain.go`, eliminating the need of changing `meta info`. `Meta info` will not be cleared, it will accumulated.